lib/puppet/provider/zfs_share/solaris.rb

```diff
@@ -0,0 +1,223 @@
+
+    if nosub
+      share_parts.insert(2, "nosub=true")
+    end
+    if anon_user_id
+      share_parts.insert(2, "anon=#{anon_user_id}")
+    end
+    if allow_ip_read
+      if allow_ip_read == '*'
+        allow_ip_full << 'ro=*'
+      elsif allow_ip_read
+        if allow_ip_read.kind_of?(String)
+          allow_ip_read = [allow_ip_read]
+        end
+        allow_ip_read_m = allow_ip_read.map { |n| "@#{n}" }
+        allow_ip_read_j = allow_ip_read_m.join(':')
+        allow_ip_full << "ro=#{allow_ip_read_j}"
+      end
+    end
+    if allow_ip_write
+      if allow_ip_write.include?('*')
+        allow_ip_full << 'rw=*'
+      elsif allow_ip_write
+        if allow_ip_write.kind_of?(String)
+          allow_ip_write = [allow_ip_write]
+        end
+        allow_ip_write_m = allow_ip_write.map { |n| "@#{n}" }
+        allow_ip_write_j = allow_ip_write_m.join(':')
+        allow_ip_full << "rw=#{allow_ip_write_j}"
+      end
+    end
+
+    if security.include?('sys')
+      share_parts << "sec=sys,#{allow_ip_full.join(',')}"
+    end
+    if security.include?('default')
+      share_parts << "sec=default,#{allow_ip_full.join(',')}"
+    end
+    if security.include?('none')
+      share_parts << "sec=none,#{allow_ip_full.join(',')}"
+    end
+
+    if protocol
+      if protocol.kind_of?(String)
+        protocol = [protocol]
+      end
+      if protocol.include?('nfs')
+        share_parts.insert(2, 'prot=nfs')
+      end
+      if protocol.include?('smb')
+        if guestok
+          guestok = 'guestok=true'
+        end
+        share_parts << ['prot=smb', guestok, allow_ip_full].compact.join(',')
+      end
+    end
+
+    share_full = share_parts.join(',')
+    return share_full
+  end
+
+  def return_cmd(cmd)
+    # "Return the output of a command as a list. Accepts a list of arguments as input."
+    output = []
+    IO.popen(cmd).each do |l|
+      output << l.chomp
+    end
+    return output
+  end
+
+  def check_for_stale_data
+    # "Check for data that present in the share on disk, but not configured to be set by the provider.
+    # When stale data is present remove it."
+    name           = share_path()
+    get_share      = ['zfs', 'get', 'share', name]
+    share          = return_cmd(get_share)
+    if share
+      share          = share[-1].split()[2]
+      share_split    = share.split(',')
+      security       = @resource[:security]
+      protocol       = @resource[:protocol]
+      allow_ip_read  = @resource[:allow_ip_read]
+      allow_ip_write = @resource[:allow_ip_write]
+      anon_user_id   = @resource[:anon_user_id]
+      nosub          = @resource[:nosub]
+      types          = ['none', 'sys', 'default']
+      if security.kind_of?(String)
+        security = [security]
+      end
+      types.reject!{ |x| security.include? x}
+      types.each do |t|
+        if share[/sec=#{t}/]
+          IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},sec=#{t}", name])
```

Comment:
I'm pretty sure this will leak open file descriptors until the GC runs. I would just add `{}` at the end to make sure it's closed. Like so:

~~~ ruby
IO.popen(['zfs', 'set', '-c', "share=#{share_split[0, 3].join(',')},sec=#{t}", name]) {}
~~~